tech_soln/real_time_dashboards.rst

```diff
@@ -0,0 +1,422 @@
+using the other replicas.
+
+.. NOTE::
+
+  In Citus Cloud you must use a replication factor of 1 (instead of the 2 used here). As
+  Citus Cloud uses `streaming replication
+  <https://www.postgresql.org/docs/current/static/warm-standby.html>`_ to achieve high
+  availability maintaining shard replicas would be redundant.
+
+With this, the system is ready to accept data and serve queries! We've provided `a data
+ingest script
+<https://github.com/citusdata/realtime-dashboards-resources/blob/master/ingest_example_data.sql>`_
+you can run to generate example data. Once you've ingested data, you can run dashboard
+queries such as:
+
+.. code-block:: sql
+
+  SELECT
+    date_trunc('minute', ingest_time) as minute,
+    COUNT(1) AS request_count,
+    SUM(CASE WHEN (status_code between 200 and 299) THEN 1 ELSE 0 END) as success_count,
+    SUM(CASE WHEN (status_code between 200 and 299) THEN 0 ELSE 1 END) as error_count,
+    SUM(response_time_msec) / COUNT(1) AS average_response_time_msec
+  FROM http_request
+  WHERE zone_id = 1 AND minute > date_trunc('minute', now()) - interval '5 minutes'
+  GROUP BY minute;
+ 
+The setup described above works, but has two drawbacks:
+
+* Your HTTP analytics dashboard must go over each row every time it needs to generate a
+  graph. For example, if your clients are interested in trends over the past year, your
+  queries will aggregate every row for the past year from scratch.
+* Your storage costs will grow proportionally with the ingest rate and the length of the
+  queryable history. In practice, you may want to keep raw events for a shorter period of
+  time (one month) and look at historical graphs over a longer time period (years).
+
+Rollups
+-------
+
+You can overcome both drawbacks by rolling up the raw data into a pre-aggregated form.
+Here, we'll aggregate the raw data into a table which stores summaries of 1-minute
+intervals. In a production system, you would probably also want something like 1-hour and
+1-day intervals, these each correspond to zoom-levels in the dashboard. When the user
+wants request times for the last month the dashboard can simply read and chart the values
+for each of the last 30 days.
+
+.. code-block:: sql
+
+  CREATE TABLE http_request_1min (
+        site_id INT,
+        ingest_time TIMESTAMPTZ, -- which minute this row represents
+
+        error_count INT,
+        success_count INT,
+        request_count INT,
+        average_response_time_msec INT,
+        CHECK (request_count = error_count + success_count),
+        CHECK (ingest_time = date_trunc('minute', ingest_time)),
+  );
+  SELECT master_create_distributed_table('http_requests_1min', 'site_id', 'hash');
+  SELECT master_create_worker_shards('http_requests_1min', 16, 2);
+  
+  -- indexes aren't automatically created by Citus
+  -- this will create the index on all shards
+  CREATE INDEX ON http_requests_1min (zone_id, ingest_time);
```

Comment:
We need to give index name, otherwise citus complains: "creating index without a name on a distributed table is currently unsupported"